Implement tapserver-based modem adapter

fuzziqersoftware opened the pull request:
This change depends on #12234, which is currently waiting on review. (When that PR is merged, I'll rebase this one to make it easier to review.)

This implements the GameCube modem adapter. This implementation appears to be stable but is not perfect; it drops frames if the receive FIFO length is exceeded. This is probably due to the unimplemented interrupt mentioned in the comment in `OnReceiveBufferSizeChangedLocked`. If the tapserver end of the connection is aware of this limitation, it's easily circumvented by lowering the MTU of the link, but ideally this wouldn't be necessary.

This has been tested with a couple of different versions of Phantasy Star Online, including Episodes 1 & 2 Trial Edition. The Trial Edition is the only version of the game that supports the Modem Adapter and not the Broadband Adapter, which is what made this PR necessary in the first place.

How state should be saved and loaded is also an open question, described a bit in the comments in DoState. Probably it shouldn't do *nothing* (as it currently does), but I don't know what exactly it should do. Comments are welcome, of course.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani reviewed (changes requested):
This review is focused on the modem part. However, I'm no modem expert so I won't be of much help regarding the technical details.

ShiftaDeband commented:
I was able to get the Windows build to work on Windows 11, ARM64, by making a few changes to the initially submitted code.  You can [view what I did over here](https://github.com/ShiftaDeband/dolphin/pull/3/files), but note that it is just... not pretty or proper. I was just trying to simply get it to work. (But it _did build_ and work at least.) It looks like `reinterpret_cast` is pretty helpful for cross compatibility?

It looks like the other thing that the Windows build doesn't like is that there are two files named `TAPServer.cpp`, even if they're in folders.

```
MSB8027	
Two or more files with the name of TAPServer.cpp will produce outputs to the same location. This can lead to an incorrect build result.  The files involved are Core\HW\EXI\BBA\TAPServer.cpp, Core\HW\EXI\Modem\TAPServer.cpp.	
```

![image](https://github.com/dolphin-emu/dolphin/assets/40249870/8837ff97-89ed-4b5c-939f-e5bbc34c2df7)

Two notes. I got this error on both MacOS and Windows when the game initially starts to connect (initializing?) or after disconnecting.

![image](https://github.com/dolphin-emu/dolphin/assets/40249870/30089c10-d653-490d-8e70-86af512da43f)

Second, when the client disconnects, and some time passes (a minute?), the log goes haywire with the following error:

<img width="1246" alt="Screenshot 2024-01-28 at 1 53 20 AM" src="https://github.com/dolphin-emu/dolphin/assets/40249870/60503a77-15fd-4b89-8700-fa4e7c4d5392">

...and keeps going on until the game goes back online or the instance is stopped:

<img width="1242" alt="image" src="https://github.com/dolphin-emu/dolphin/assets/40249870/42471ba1-3423-4bda-a044-866293ac3049">

fuzziqersoftware commented on Source/Core/Core/HW/EXI/EXI_DeviceModem.h:
```diff
@@ -0,0 +1,179 @@
+  void DoState(PointerWrap& p) override;
+
+private:
+  enum Interrupt
+  {  // Used for Register::INTERRUPT_MASK and Register::PENDING_INTERRUPT_MASK
+    AT_REPLY_DATA_AVAILABLE = 0x02,
+    SEND_BUFFER_BELOW_THRESHOLD = 0x10,
+    RECEIVE_BUFFER_ABOVE_THRESHOLD = 0x20,
+    RECEIVE_BUFFER_NOT_EMPTY = 0x40,
+  };
+  enum Register
```
This was based on reverse-engineering PSO Episodes 1&2 Trial Edition. These names are just names I chose based on what the flags seem to be doing; they aren't based on any documentation.

fuzziqersoftware commented on Source/Core/Core/HW/EXI/EXI_DeviceModem.h:
```diff
@@ -0,0 +1,179 @@
+namespace ExpansionInterface
+{
+
+#define MODEM_RECV_SIZE 0x800
```
This is what the BBA does, so I copied its pattern. I assume the BBA should be changed then too?

fuzziqersoftware commented on Source/Core/Core/HW/EXI/EXI_DeviceModem.cpp:
```diff
@@ -0,0 +1,382 @@
+    std::string command = m_at_command_data.substr(0, newline_pos);
+    m_at_command_data = m_at_command_data.substr(newline_pos + 1);
+
+    if (command == "ATZ")
+    {  // Reset
+      m_network_interface->Deactivate();
+      AddATReply("OK\r");
```
It seems this is true, but it also seems the modem library only ever sends ATZ with no parameter. Regardless, I'll add support for the parameter and add command logging.

fuzziqersoftware commented:
I've updated this PR with changes based on the review comments here and in #12234, and have rebased onto main. I have not tested the Windows build, but have incorporated ShiftaDeband's changes into this PR as well, so hopefully it will work on Windows too.

sepalani commented:
@dolphin-emu-bot rebuilt

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented:
@dolphin-emu-bot rebuild

ShiftaDeband commented:
As for the log issue, it looks like `ATH` commands may not be handled? At least as far as I could tell.

[ATH0 is hook control, go on hook/disconnect](https://www.lammertbies.nl/comm/info/hayes-at-commands#at0h) (ATH1 is apparently to go off hook/connect, not sure if used.) When a client disconnects, the GameCube's modem sends this to the other end.

When handled inside the `RunAllPendingATCommands`, adding an else if statement seems to resolve the repeating log message issue mentioned earlier:

```
...
    if (command.substr(0, 3) == "ATZ")
    {
      // Reset
      m_network_interface->Deactivate();
      AddATReply("OK\r");
    }
    else if (command == "ATH0")
    {
      // Disconnect
      m_network_interface->Deactivate();
      AddATReply("OK\rNO CARRIER\r");
    }
...
```

Of course, reconnecting and everything still works fine. Although when using this, `TAPServerConnection.cpp` still logs one error. Doesn't seem to be a huge issue but wanted to at least report this.

<img width="294" alt="Screenshot 2024-02-07 at 8 31 14 PM" src="https://github.com/dolphin-emu/dolphin/assets/40249870/85c11706-87d1-4667-90cd-384044d4436c">

Anyway, I've tested this with several revisions of PSO from different regions and haven't had any issues. This is a great addition, again, thank you.

fuzziqersoftware commented:
Thanks for investigating; I've added a case to handle ATH0 as you recommended. I believe the remaining error is because we close the socket before terminating the read thread, so in some cases it will try to read from a closed socket. I've switched the order, which should prevent this in the future, even though it is harmless.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented on Source/Core/Core/HW/EXI/EXI_DeviceModem.cpp:
```diff
@@ -0,0 +1,396 @@
+    bool should_update_interrupts = false;
+    for (; size; size--)
+    {
+      if (reg_num >= m_regs.size())
+      {
+        break;
+      }
```
Can't this condition be part of the `for`'s condition _(e.g. `size && reg_num < m_regs.size()`)_?

sepalani commented on Source/Core/Core/HW/EXI/EXI_DeviceModem.cpp:
```diff
@@ -0,0 +1,396 @@
+  {
+    // AT command buffer
+    const std::size_t bytes_to_copy = std::min<std::size_t>(size, m_at_reply_data.size());
+    memcpy(data, m_at_reply_data.data(), bytes_to_copy);
```
`std::memcpy`

sepalani commented on Source/Core/Core/HW/EXI/EXI_DeviceModem.cpp:
```diff
@@ -0,0 +1,396 @@
+    // Packet receive buffer
+    std::lock_guard<std::mutex> g(m_receive_buffer_lock);
+    const std::size_t bytes_to_copy = std::min<std::size_t>(size, m_receive_buffer.size());
+    memcpy(data, m_receive_buffer.data(), bytes_to_copy);
```
`std::memcpy`

sepalani reviewed (changes requested):
Here is another pass.

BTW, can't the classes inherit `TAPServerConnection` so they don't have to reimplement its methods?

fuzziqersoftware commented on Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp:
```diff
@@ -0,0 +1,351 @@
+using ws_ssize_t = int;
+#else
+#define closesocket close
+using ws_ssize_t = ssize_t;
```
std::ssize_t doesn't exist.

sepalani replied:
My bad I completely forgot that it wasn't defined in C++'s `cstdlib` but in C's `sys/types.h`.

sepalani commented:
@dolphin-emu-bot rebuild

sepalani commented on Source/Core/Core/HW/EXI/BBA/TAPServerConnection.h:
```diff
@@ -0,0 +1,56 @@
+  };
+
+  const std::string m_destination;
+  const std::function<void(std::string&&)> m_recv_cb;
```
`const RecvCallback`

sepalani commented on Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp:
```diff
@@ -0,0 +1,351 @@
+        if (frame_bytes_expected > m_max_frame_size)
+        {
+          ERROR_LOG_FMT(SP1, "Packet is too large ({} bytes); dropping it", frame_bytes_expected);
+          read_state = ReadState::SKIP;
+        }
+        else
+        {
```
You think so? There are several places in your code where you did `return`/`break` on error/edge cases before my review.

TBH, I don't have a strong stance about it either, so feel free to disregard that point if you don't feel comfortable changing it.

sepalani reviewed (changes requested):
Thanks a lot for your time. Beside that, the code LGTM.

Sadly, I have no way to test the PR features.

Matt-Swift commented:
I've been playing around with my own PSO newserv instance on the tapserver modem and BBA.

The only issue I've had with the tapserver modem is related to newserv not sending enough information fast enough, although this does mean I haven't been able to do any extensive testing with it, but otherwise it is quite stable.

However, there does seem to be an issue with the tapserver BBA where Dolphin will seemingly cut the connection to the PSO server instance randomly. As the PSO client didn't send a disconnect packet or receive any sort of disconnection instruction from the server, the game will be softlocked until the player quits the game themselves (or the client responds to having received no packets from the server after 15 minutes).

The newserv packet logs don't indicate anything amiss, as it happens during innocent scenarios, and the client just times out when the newserv instance doesn't receive any responses.

After this, trying to go online again will immediately crash Dolphin. 

I managed to get a backtrace under gdb with Dolphin debug of this crash, although it's not quite under the same circumstances:

[tapserver-bba-crash.txt](https://github.com/dolphin-emu/dolphin/files/14406347/tapserver-bba-crash.txt)

This seems 100% reproducible by running Dolphin Debug under gdb, loading game GPOE8P (Phantasy Star Online Episode I & II Plus US), going online, quitting game, then going online again. If you want to test on a public newserv instance rather than a local one, `217.160.146.96:9501` will accept tapserver BBA, and `217.160.146.96:9502` will accept tapserver modem.

sepalani commented on Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp:
```diff
@@ -0,0 +1,356 @@
+using ws_ssize_t = ssize_t;
+#endif
+
+#ifdef __LINUX__
```
This should be `__linux__` (in lowercase) instead. It should fix the debian sigpipe crash.

sepalani commented on Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp:
```diff
@@ -0,0 +1,356 @@
+    timeval timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_usec = 50000;
+    if (select(m_fd + 1, &rfds, nullptr, nullptr, &timeout) <= 0)
+      continue;
+
```
This might silent errors, you can continue on `0`/`timeout` but if an error occurs (< 0) it should be handled and logged properly.

Matt-Swift commented:
With the new build, Dolphin no longer crashes when trying to connect online after the tapserver BBA bug kicks in (it does still fail to reconnect, however). In regular usage, this bug seems to occur after a few hours randomly, and sometimes it will also occur after a regular disconnection after some play time too.

On the debug build under gdb (and not otherwise, whether using debug only or gdb only), it happens every single time you connect online, then disconnect, meaning this is quickly reproducible.

I'm not sure how to get any sort of useful trace since the program isn't crashing, but I did decide to look at what was happening on the network with wireshark from the first connection, followed by the second (broken) connection:

[normal-connection.txt](https://github.com/dolphin-emu/dolphin/files/14434078/normal-connection.txt)
[broken-connect.txt](https://github.com/dolphin-emu/dolphin/files/14434081/broken-connect.txt)

sepalani commented on Source/Core/Core/HW/EXI/BBA/TAPServerConnection.cpp:
```diff
@@ -0,0 +1,362 @@
+    timeout.tv_sec = 0;
+    timeout.tv_usec = 50000;
+    int select_res = select(m_fd + 1, &rfds, nullptr, nullptr, &timeout);
+    if (select_res < 0)
+    {
+      ERROR_LOG_FMT(SP1, "Can\'t poll tapserver fd: {}", Common::StrNetworkError());
+      break;
```
You shouldn't break of the `while` loop, otherwise the `ReadThreadHandler` will finish and can't be used anymore.

If there is an error, this might suggest that the connection was closed/interrupted with the TAP server. You might want to create a `Reconnect` method which will close the failing socket and try to reconnect (maybe x times?) to the destination. If the reconnect method fails, you might want to close the socket and set the fd to -1.

Moreover, you might want to add the following:
 - Check that `m_read_enabled` is set before handling the socket:
```c++
  while (!m_read_shutdown.IsSet())
  {
    if (!m_read_enabled.IsSet())
      continue;
```
 - Then, reconnect to the TAP server if needed in `RecvStart`:
```c++
if (m_read_enabled.IsSet() && !IsActivated())
  ReconnectToDestination(m_destination);
```

fuzziqersoftware replied:
There is a good reason why we don't attempt to reconnect. The tapserver connection includes state on the remote end of the connection, so reconnecting isn't as simple as just opening the TCP connection again; we would also need to tell the remote end what its virtual network state should be, and it would have to reopen the downstream connections that were part of the original virtual network. Those connections may also have had server-side state of their own which was lost (in the case of PSO, they definitely would have), so reconnecting after a disconnection simply isn't feasible even if we were to reconstruct the remote tapserver's state perfectly.

sepalani replied:
This still needs no be addressed as by doing nothing the game will still believe it is connected while it isn't. My comment regarding `m_read_enabled` and not breaking from the while loop still apply _(which might be responsible for the reconnection issue mentioned by @Matt-Swift)_.

Some questions regarding the error handling:
- What does the TAP server do when there is a client disconnection? _(i.e. does it notify the connected peers like the PSO server?)_
- What does the TAP server do when the client logs out and disconnects gracefully? _(i.e. does it close the TAP server connection and does the TAP server disconnect the client from its peers?)_
- What does the (PSO) game server do when a client disconnects? _(e.g. from a direct connection or a TAP server)_
- What does the (PSO) game server do when a client is trying to connect twice? _(i.e. would both be successful, would the first connection be disconnected or would the last connection raise an error?)_

Is the TAP server used open-source? Unless I'm mistaken, if it acts just like a regular TAP, trying to reconnect to it when there was a socket error (which probably means we got disconnected from the TAP server) shouldn't be that harmful. It either retained its previous state (which might be fine) or it doesn't and will try to create a new one (which should be fine as long as we notified the game that we were disconnected).

fuzziqersoftware replied:
[tapserver](https://github.com/fuzziqersoftware/tapserver) is a program that creates a "fake ethernet" interface when an inbound connection is received, and uses that to forward and receive traffic through the kernel's network stack. it does not act like a regular tap interface; it's designed to be similar, but there are some semantic differences. Notably, when the tapserver connection is closed, the associated network interface is deleted, which causes the kernel to close any TCP connections that were connected through it.

[newserv](https://github.com/fuzziqersoftware/newserv) implements the same protocol as tapserver, but implements its own Ethernet, PPP, IP, and TCP stacks internally, which does a better job of handling some of the quirks of PSO than just using tapserver and the modern kernel's network stack directly. In both projects, all TCP connections tunneled through the tapserver connection are closed when the tapserver connection is closed.

To answer your specific questions,
* When the client disconnects either normally or unexpectedly, tapserver does disconnect upstream TCP connections.
* When a client disconnects from the game server, the server removes the client from the lobby it was in (if any), notifies other clients in that lobby, deletes any game state associated with the player, and saves their character files if necessary.
* When a client connects twice simultaneously, the server accepts both connections, and these connections do not share any state (they may use different characters and be in different lobbies/games). newserv, at least, is intentionally designed to allow multiple logins on the same account at the same time, thought most other PSO servers don't allow this.

What do the other network interfaces do when their underlying links go down? If an actual tap interface (not tapserver) becomes unavailable, or an XLink socket can't receive for some reason, it looks like those BBA types would just loop forever trying to receive and logging error messages. Furthermore, it looks like the BBA NetworkInterface abstraction doesn't have any way to notify CEXIETHERNET that the link is down due to an external failure. This is not to say we shouldn't handle that case in the modem adapter, but the point of this PR is to bring its implementation up to speed with how the BBA currently works, and I believe that has been done. Figuring out how to notify the game that the interface is down due to an external failure will require significantly more reverse-engineering work, and I am not willing to spend an arbitrary amount of time making this perfect when it's already at parity with the other existing implementations.

Matt-Swift replied:
I admit that while this issue isn't perfect, I would be quite interested in seeing this implemented sooner rather than later, as ironically, the current BBA HLE that is implemented in Dolphin is not actually usable with Phantasy Star Online Episode I & II as the game will crash while online after 30 minutes to hour without fail (this is a separate issue that needs reporting if it hasn't already); this tapserver implementation on the other hand works flawlessly, and having to quickly restart the game in a non-critical spot is a major improvement over the current experience.

The reported issue of randomly disconnecting may have just been a fluke from my end, as a friend and I have been playing using this version of Dolphin without any issues recently on my own public newserv instance, but most importantly without any client freezes.

The most crucial addition of course though as well is the modem adapter, making Episode I & II Trial Edition accessible to many more players, and from my cursory testing with the modem adapter, there doesn't seem to be any unique issues with it.

sepalani replied:
I do get your point but the current implementation isn't on par with the Win32 TAP implementation nor Unix implementation.

@fuzziqersoftware 
That's why I suggested to, at least, address the following issues:
- `m_read_enabled` needs to be checked before handling reads.
- Not exiting the while loop if there is an error on `select`, you can use `continue` instead _(like what the TAP Unix code uses to do)_ and add an error log.

@Matt-Swift 
Regarding the BBA HLE issue, there is an issue for it:
 - https://bugs.dolphin-emu.org/issues/13032

However, the lack of reproducibility, logs and the fact I don't own the game, that makes it very difficult to diagnose the root cause. My best attempt to fix the issue is this PR:
 - https://github.com/dolphin-emu/dolphin/pull/12560

As I don't own the game, I'm unsure if it solves the issue. Regardless, I also believe that tunneling the traffic to a TAP server is probably faster than emulating sending/receiving frames.

fuzziqersoftware replied:
I'm still not convinced that repeating select() after it has failed once is the correct thing to do (here or in the BBA implementation), but this PR has taken long enough already and I'm tired, so I've changed it as you suggested.

sepalani commented:
@dolphin-emu-bot rebuild

Shoegzer commented:
@fuzziqersoftware I'd like to help test this with PSO, however the modem adapter option doesn't appear to be exposed in the UI with a build made after this PR was merged.  Currently shown under SP1: Nothing, Dummy and the three BBA options.  Sorry to ask if it's obvious but, where exactly can this be found?  This could be the only working method for PSO as the game doesn't work with the BBA HLE, even after testing #12560 (as previously stated, freezes at unknown intervals causing loss of game data).

Matt-Swift commented:
It's in the same location, if you've built from dolphin-emu:master, it should just be visible under SP1. If for some reason it doesn't appear (I have not tested the master branch), build from [fuzziqersoftware's fork](https://github.com/fuzziqersoftware/dolphin) and checkout the tapserver-modem-adapter branch and build from there.

If you're not on Linux, if you download one of the Dolphin versions from [here](https://dolphin-emu.org/download/dev/b510ac89a31febc07b6764d3f5d704b1e1a5fe8b/), this will expose the tapserver BBA in SP1 for sure, as other users have already used this on my public newserv instance.

Note that this BBA option only works on servers which support the tapserver interface, which at time of writing, is any [newserv](https://github.com/fuzziqersoftware/newserv) instance, and the only known public one would be at [ragol.org](https://ragol.org). To test it out there, put `ragol.org:9501` in the configuration box for the tapserver BBA (or `ragol.org:9502` for the modem) and it will just connect.

Shoegzer commented:
Thanks, I'll try that. I'm on Linux Mint 21.3 and I built from current master but the adapter doesn't show.  I'll try the fork when I get a chance though, and if the interface shows there I'll test against ragol.
 
Incidentally, [this](https://ragol.org/gc-connect) section of the ragol website directs users to Dolphin 5.0-16838 as "Future versions have constant FSOD issues." While I haven't tested that build myself, notaloop reported [here](https://bugs.dolphin-emu.org/issues/13032#note-22) that it too exhibits the freezing issue.  If the modem adapter fixes that problem though, whomever maintains ragol (hopefully they're reading this) should update it to attract more players.  

It does leave me wondering though, whether this issue is triggered by a specific event from another logged-in user and not necessarily random (therefore a server with more players would be more likely to trigger the issue, and it may not manifest as often on a less active server).  I was thinking schtserv transitioned to newserv but I suppose it hasn't.  That instance seems to attract the most players, partially due to the fact that the HLE BBA defaults to it.  My theory is just a wild guess and probably wrong, but if not, schthack would probably make the best testing ground.

Matt-Swift commented:
PSO itself can cause the issue, but it is probably true that 16838 also has the issue, just to a lesser degree.

Anyway, a closed git forge PR is probably not the best place to discuss these things, but here's the details to the above:

- I am the owner of the ragol.org PSO newserv instance, so the page has been updated. As I'm sure you can appreciate, this only got merged yesterday, although there is a news post about it.
- SCHTServ will most likely never "transition to newserv" as it uses its own closed-source codebase written in Delphi. They can choose to host their own tapserver instance to accept tapserver BBA requests if they so choose. [tapserver can be found here](https://github.com/fuzziqersoftware/tapserver).
- The issue is not triggered by a specific event from a logged-in user, nor by the server specifically. This issue happens on SCHTServ, Sylverant, and newserv regardless of how many players are present - it can even happen alone. We have played many PSO games at Ragol, and FSODs are extremely common on players on the latest Dolphin, but have never occurred to players using the tapserver BBA.

Shoegzer commented:
Thanks, that's quite interesting.

* Indeed, given the fast pace of the new developments it's understandable that the site wasn't updated.  At the time I didn't know if the changes would be noticed anytime soon, but I'm glad to know you're the owner!
* I had no idea the schtserv backend was closed-source but that's good to know.  And I agree it doesn't sound likely they would transition away from the codebase they've developed and maintained for so many years now.
* I didn't realize the FSOD happened when a single user was logged in but that's good to know.

I'm not sure why the new tapserver options weren't showing earlier, but they do now on a build I just made from master (so you can remove the fork reference for linux users on your page, and replace it with the requirement for Dolphin 5.0-21253 or higher).  I configured the BBA tapserver for ragol.org:9501 and it worked without any other changes, with one issue: on first attempt to load a mission it got to 21% and then I got the FSOD with buzzing noise.  On second and third attempts it didn't do that though, and I've been on testing for a while with no issues.  I'm guessing it was a one-time issue then, since it happened during a load event, and I'm hoping not anything necessarily wrong with the tapserver implementation.

You're right though, a closed PR isn't exactly the best place to discuss issues that aren't related to the tapserver itself, though I can provide more detail on the above if anyone needs it.

Pokechu22 commented on Source/Core/Core/HW/EXI/EXI_DeviceModem.cpp:
```diff
@@ -0,0 +1,398 @@
+  return true;
+}
+
+void CEXIModem::SetCS(int cs)
+{
+  m_transfer_descriptor = INVALID_TRANSFER_DESCRIPTOR;
+}
```
Shouldn't this only apply if `cs` is true, similar to this?

https://github.com/dolphin-emu/dolphin/blob/a5e410df11b83646f21941caba98674d807daa7e/Source/Core/Core/HW/EXI/EXI_DeviceEthernet.cpp#L106-L113

(I noticed this with b5a6fc11bc362da7d7b62a352a62dca9922f5963 on #12680, as there I changed how `SetCS` works and used `if (!was_selected && is_selected)` for setting it to invalid. But I'm not 100% sure if that actually makes sense or not.)
